spring-xsuaa/src/main/java/com/sap/cloud/security/xsuaa/extractor/IasXsuaaExchangeBroker.java

```diff
@@ -1,55 +1,73 @@
 package com.sap.cloud.security.xsuaa.extractor;
 
+import com.nimbusds.jwt.JWT;
+import com.nimbusds.jwt.JWTParser;
 import com.sap.cloud.security.token.Token;
-import com.sap.cloud.security.token.TokenClaims;
-import com.sap.cloud.security.xsuaa.jwt.Base64JwtDecoder;
+import com.sap.cloud.security.xsuaa.XsuaaServiceConfiguration;
+import com.sap.cloud.security.xsuaa.client.ClientCredentials;
+import com.sap.cloud.security.xsuaa.client.OAuth2TokenService;
+import com.sap.cloud.security.xsuaa.client.XsuaaDefaultEndpoints;
+import com.sap.cloud.security.xsuaa.client.XsuaaOAuth2TokenService;
 import com.sap.cloud.security.xsuaa.tokenflows.TokenFlowException;
 import com.sap.cloud.security.xsuaa.tokenflows.XsuaaTokenFlows;
-import org.json.JSONException;
-import org.json.JSONObject;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+import org.springframework.security.oauth2.jwt.Jwt;
+import org.springframework.security.oauth2.server.resource.InvalidBearerTokenException;
+import org.springframework.security.oauth2.server.resource.web.BearerTokenResolver;
+import org.springframework.web.client.RestTemplate;
 
 import javax.annotation.Nullable;
-
-import static com.sap.cloud.security.token.TokenClaims.XSUAA.EXTERNAL_ATTRIBUTE_ENHANCER;
+import javax.servlet.http.HttpServletRequest;
+import java.text.ParseException;
 
 /**
  * IAS token and XSUAA token exchange and resolution class. Can be used to
  * distinguish between IAS and XSUAA tokens. Controls token exchange between IAS
  * and XSUAA by using IAS_XSUAA_XCHANGE_ENABLED environment variable flag
  */
-public class IasXsuaaExchangeBroker {
+public class IasXsuaaExchangeBroker implements BearerTokenResolver {
 
 	private static final Logger logger = LoggerFactory.getLogger(IasXsuaaExchangeBroker.class);
 
 	private final XsuaaTokenFlows xsuaaTokenFlows;
-	private final boolean isIasXsuaaXchangeEnabled;
-	private static final String IAS_XSUAA_ENABLED = "IAS_XSUAA_XCHANGE_ENABLED";
+	private static final String AUTH_HEADER = "Authorization";
 
 	public IasXsuaaExchangeBroker(XsuaaTokenFlows xsuaaTokenFlows) {
 		this.xsuaaTokenFlows = xsuaaTokenFlows;
-		this.isIasXsuaaXchangeEnabled = resolveIasToXsuaaEnabledFlag();
 	}
 
-	/**
-	 * Verifies if the provided token is Xsuaa token
-	 *
-	 * @param encodedJwtToken
-	 *            Encoded token to be checked
-	 * @return true if provided token is a XSUAA token
-	 */
-	public boolean isXsuaaToken(String encodedJwtToken) {
-		String claims = Base64JwtDecoder.getInstance().decode(encodedJwtToken).getPayload();
+	public IasXsuaaExchangeBroker(XsuaaServiceConfiguration configuration) {
+		this.xsuaaTokenFlows = new XsuaaTokenFlows(
```

Comment:
- you can reuse the 3rd constructor by just providing the tokenService
- i would prefer as default tokenService : new DefaultOAuth2TokenService()